include/engine/api/route_api.hpp

```diff
@@ -215,43 +228,54 @@ class RouteAPI : public BaseAPI
 
         std::vector<util::json::Object> annotations;
 
-        if (parameters.annotations)
+        if (parameters.annotations_type != RouteParameters::AnnotationsType::None)
         {
             for (const auto idx : util::irange<std::size_t>(0UL, leg_geometries.size()))
             {
-                util::json::Array durations;
-                util::json::Array distances;
-                util::json::Array weights;
-                util::json::Array nodes;
-                util::json::Array datasources;
                 auto &leg_geometry = leg_geometries[idx];
-
-                durations.values.reserve(leg_geometry.annotations.size());
-                distances.values.reserve(leg_geometry.annotations.size());
-                weights.values.reserve(leg_geometry.annotations.size());
-                nodes.values.reserve(leg_geometry.osm_node_ids.size());
-                datasources.values.reserve(leg_geometry.annotations.size());
-
-                std::for_each(leg_geometry.annotations.begin(),
-                              leg_geometry.annotations.end(),
-                              [this, &durations, &distances, &weights, &datasources](
-                                  const guidance::LegGeometry::Annotation &step) {
-                                  durations.values.push_back(step.duration);
-                                  distances.values.push_back(step.distance);
-                                  weights.values.push_back(step.weight);
-                                  datasources.values.push_back(step.datasource);
-                              });
-                std::for_each(leg_geometry.osm_node_ids.begin(),
-                              leg_geometry.osm_node_ids.end(),
-                              [this, &nodes](const OSMNodeID &node_id) {
-                                  nodes.values.push_back(static_cast<std::uint64_t>(node_id));
-                              });
                 util::json::Object annotation;
-                annotation.values["distance"] = std::move(distances);
-                annotation.values["duration"] = std::move(durations);
-                annotation.values["weight"] = std::move(weights);
-                annotation.values["nodes"] = std::move(nodes);
-                annotation.values["datasources"] = std::move(datasources);
+
+                if (parameters.annotations_type & RouteParameters::AnnotationsType::Duration)
+                {
+                    annotation.values["duration"] =
+                        GetAnnotations(leg_geometry,
+                                       std::bind(&guidance::LegGeometry::Annotation::duration,
+                                                 std::placeholders::_1));
```

Comment:
Instead of bind you can simply use a lambda, which is probably easier to read and does not have all the weird quirks bind has (see https://www.youtube.com/watch?v=zt7ThwVfap0). E.g.

```cpp
[](const auto& v) { return v.distance };
```